Move the scheduling of array task into the FifoScheduler

WilliamMcCumstie opened the pull request:
Previously it was assumed that each `ARRAY_TASK` would acquire a single node from within its `ARRAY_JOB` "virtual allocation". NOTE: the `ARRAY_JOB` never actually received an allocation, but conceptually the above assumption still holds.

However this assumption was incorrect as each `ARRAY_TASK` must be assigned the minimum number of nodes according to its `ARRAY_JOB`. This creates a scheduling problem which requires more formal integration with the `FifoScheduler`.

The ultimate goal is to remove the use of the `TaskRegistry` as it does not have a clear distinction in purpose from scheduler. The concept of a "group of jobs" will take its place. Note that fully removing the `TaskRegistry` is out of scope as it has implications on job cancellation.

Instead each `Job` now has a "group of jobs" which can be enumerated over. The groupings are as follows:
* `JOB`: Standard batch jobs have a group containing themself,
* `ARRAY_JOB`: Have a group containing all of their `ARRAY_TASK`s but not themself, and
* `ARRAY_TASK`: Are the degenerative case and contain an empty group as they can not be directly scheduled

When a job is added to the `FifoScheduler` it is stored via it's group. This allows the scheduler to expand the group on the fly to generate the current state of the queue. It will attempt to run all of the jobs within group before moving onto the next group of jobs. The previous "pass over" ability has been removed as it is conceptually incompatible with FIFO. Other schedulers may handle the groupings differently.

Other changes include the removal of the `TaskSerializer` as its no longer required. Each "task" now appears in the scheduler `queue` as a consequence of the above groupings.

WilliamMcCumstie committed:
Add an enumerator method to jobs

This will be used by the scheduler to abstract the differences
between the JOB and ARRAY_JOB type

WilliamMcCumstie committed:
Start refactoring FifoScheduler to handle array jobs

The FifoScheduler will eventually replace the `task_registry`. This
requires it to take on grouping concepts. This requires an extensive
internal refactor which will be done over the following commits

The failing specs will be re-instated one by one

WilliamMcCumstie committed:
Fix the trivial case empty FifoScheduler case

WilliamMcCumstie committed:
Reimplement allocate_job to work with the enumerator

This removes all the specific details about ARRAY_JOB vs JOB
from the scheduler

WilliamMcCumstie committed:
Remove the uses of scheduler.queue from the jobs

The queue method does not behave nicely due to the nature of
array jobs. It will need independent testing

As such the other specs should not rely on it as part of their
setup

WilliamMcCumstie committed:
Fix the remaining scheduler spec

NOTE: This does not correctly clean up the scheduler, but the
remaining spec does not correctly test this fact

WilliamMcCumstie committed:
Reimplement the queue method for batch jobs

WilliamMcCumstie committed:
Include the array tasks in the queue

WilliamMcCumstie committed:
Include the main array job if it has unstarted tasks

WilliamMcCumstie committed:
Do not queue the job until it has data

It is okay if the data exists before the job is queued, but not the
other way round

WilliamMcCumstie committed:
Fix the remove_job method for batch jobs

WilliamMcCumstie committed:
Allow tasks to be partially removed

WilliamMcCumstie committed:
Remove the ARRAY_JOB when its last task finishes

WilliamMcCumstie committed:
Use the scheduler to fetch the active tasks

This removes the use of the `task_registry` in the serializer

WilliamMcCumstie committed:
Remove task serialization and infer array job ids

All if the indexing and task stuff is so the client can work out
how the tasks fit in with the jobs

However now all the tasks exist within the scheduler, this isn't
required

For ease client side, the ARRAY_JOB and ARRAY_TASK ids are now
determined server side

WilliamMcCumstie committed:
Pass the minimum number of nodes to each task

WilliamMcCumstie committed:
Reinstate the jobs_controller_spec

benarmston reviewed (changes requested):
Cleaning up an array job seems broken.

```
$ ./bin/scheduler batch --array 1 --nodes 1 tmp/simple-array.sh  this argument
Submitted batch job 3d6481e1-3e40-477d-b859-b9f2d04e7937[-1]
$ ./bin/scheduler queue
┌─────────────────────────────────────────┬───────────┬────────┬──────┬───────────┬──────┬───────┬──────────────────┐
│ JOBID                                   │ PARTITION │ NAME   │ USER │ ST        │ TIME │ NODES │ NODELIST(REASON) │
├─────────────────────────────────────────┼───────────┼────────┼──────┼───────────┼──────┼───────┼──────────────────┤
│ 3d6481e1-3e40-477d-b859-b9f2d04e7937[1] │ all       │ (none) │ TBD  │ COMPLETED │ TBD  │ 1     │ node01           │
└─────────────────────────────────────────┴───────────┴────────┴──────┴───────────┴──────┴───────┴──────────────────┘
```

The task sits in the queue with state `COMPLETED` indefinitely.

I'm running this on branches: cli=`dev/remove-tasks`, controller=`fix/array-task-allocation`, daemon=`master`.

---

The general idea you've taken looks good.  When we add job persistence we may need to reconsider the use of an enumerator, as I'm not sure how easy it is to save/restore their current state.

I'll take another look when you've fixed the above issue.

WilliamMcCumstie commented:
In regards to the `COMPLETED` task sticking around, I think it's a red hearing. I couldn't replicate the issue:

```
[root@master ~]# flight scheduler batch /tmp/long --array 1 --nodes 1  this argument
Submitted batch job e823d168-0619-4e0e-bebb-e71892dced27[-1]
[root@master ~]# flight scheduler queue
┌─────────────────────────────────────────┬───────────┬────────┬──────┬─────────┬──────┬───────┬──────────────────┐
│ JOBID                                   │ PARTITION │ NAME   │ USER │ ST      │ TIME │ NODES │ NODELIST(REASON) │
├─────────────────────────────────────────┼───────────┼────────┼──────┼─────────┼──────┼───────┼──────────────────┤
│ e823d168-0619-4e0e-bebb-e71892dced27[1] │ all       │ (none) │ TBD  │ RUNNING │ TBD  │ 1     │ node01           │
└─────────────────────────────────────────┴───────────┴────────┴──────┴─────────┴──────┴───────┴──────────────────┘
[root@master ~]# flight scheduler queue
.. no output ..
```

This is using the slightly more modern builds, but I doubt this is the cause of the difference:
```
[root@master ~]# yum info flight-scheduler-controller | grep Version
Version     : 0.3.0
[root@master ~]# yum info flight-scheduler | grep Version
Version     : 0.3.0

# On the daemon
[root@master ~]# yum info flight-scheduler-daemon | grep Version
Version     : 0.2.0
```

My guess is the `NODE_DEALLOCATED` message wasn't sent from the `daemon` or it was missed by the `controller`. You will need to check the logs to determine if this is the case. A separate tracking card has been opened as I don't believe it's related to this PR.

WilliamMcCumstie commented:
I get the same output as above when using the following git versions:
```
flight-scheduler: bc0c030608fe5f3651211475b4c790406718422a
flight-scheduler-scheduler: d6ee111827b762952f2842ff034ef485d993e332
flight-scheduler-daemon: b850b2e5d866fffcd80cf6b61046ac9fda62b28e
```

EDIT: The git sha sum was initially incorrect for `flight-scheduler`

WilliamMcCumstie commented:
From external discussions it appears a `job_step` is blocking the scheduler from removing the job. The following runs with this demo script:

```
# cat /tmp/step-long 
#!/bin/bash
value=$(shuf -i 10-30 -n 1)
${flight_ROOT}/bin/flight scheduler run --jobid ${JOB_ID} sleep "$value"
```

Then client side I get this output:

```
[root@master scheduler]# bin/scheduler batch /tmp/step-long -N 2 -a 1
Submitted batch job 99d896c0-b9e2-4213-84bc-16fb0a42a773[-1]
[root@master scheduler]# bin/scheduler queue
┌─────────────────────────────────────────┬───────────┬────────┬──────┬───────────┬──────┬───────┬──────────────────┐
│ JOBID                                   │ PARTITION │ NAME   │ USER │ ST        │ TIME │ NODES │ NODELIST(REASON) │
├─────────────────────────────────────────┼───────────┼────────┼──────┼───────────┼──────┼───────┼──────────────────┤
│ 99d896c0-b9e2-4213-84bc-16fb0a42a773[1] │ all       │ (none) │ TBD  │ COMPLETED │ TBD  │ 2     │ node01,node02    │
└─────────────────────────────────────────┴───────────┴────────┴──────┴───────────┴──────┴───────┴──────────────────┘
# .. pause for some amount of time ..
[root@master scheduler]# bin/scheduler queue
.. no output ..
```

It appears a job step is blocking the daemon from sending `NODE_DEALLOCATED` which causes the job to stick around in the scheduler with a `COMPLETED` status. In my case the job step exists normally and releases the node as expected. However if the job step permanently hangs, then it will never release the node (unless a cancel message is sent?).

benarmston commented:
@william rebasing this PR and the CLI PR on their respective `master` branches has mostly resolved the issue.

There is still the possibility that a job step can hang indefinitely but that will be addressed when we consider adding job time limits.

benarmston commented:
It looks like cancelling running array jobs is broken.

```
ben@glitnir 15:19:10 flight-scheduler (dev/remove-tasks *$%=)
$ cat tmp/super-simple-array.sh 
#!/bin/bash

#SBATCH -N 1

echo "sleeping"
sleep 120
echo "done sleeping"
ben@glitnir 15:19:14 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler batch --array 1,2  --nodes 1 tmp/super-simple-array.sh
Submitted batch job 19506270-22ca-4199-9204-6d43222319c5[-2]
ben@glitnir 15:19:21 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler queue
┌─────────────────────────────────────────┬───────────┬────────┬──────┬─────────┬──────┬───────┬──────────────────┐
│ JOBID                                   │ PARTITION │ NAME   │ USER │ ST      │ TIME │ NODES │ NODELIST(REASON) │
├─────────────────────────────────────────┼───────────┼────────┼──────┼─────────┼──────┼───────┼──────────────────┤
│ 19506270-22ca-4199-9204-6d43222319c5[1] │ all       │ (none) │ TBD  │ RUNNING │ TBD  │ 1     │ node01           │
│ 19506270-22ca-4199-9204-6d43222319c5[2] │ all       │ (none) │ TBD  │ RUNNING │ TBD  │ 1     │ node02           │
└─────────────────────────────────────────┴───────────┴────────┴──────┴─────────┴──────┴───────┴──────────────────┘
ben@glitnir 15:19:26 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler cancel 19506270-22ca-4199-9204-6d43222319c5
bin/scheduler: The API returned a 404 error status and this content:
  {
    "errors": [
      {
        "id": "c05a44f8-1845-43f3-861b-2abad604b417",
        "title": "Not Found Error",
        "detail": "Resource '19506270-22ca-4199-9204-6d43222319c5' not found",
        "status": "404"
      }
    ]
  }
ben@glitnir 15:19:32 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler cancel '19506270-22ca-4199-9204-6d43222319c5[1]'
bin/scheduler: bad URI(is not URI?): "/v0/jobs/19506270-22ca-4199-9204-6d43222319c5[1]"
ben@glitnir 15:19:38 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler cancel '19506270-22ca-4199-9204-6d43222319c5%5B1%5D'
bin/scheduler: The API returned a 404 error status and this content:
  {
    "errors": [
      {
        "id": "22856981-4bd1-4abf-b5f8-8c20f5bdf87b",
        "title": "Not Found",
        "detail": "DELETE /jobs/19506270-22ca-4199-9204-6d43222319c5%5B1%5D",
        "status": "404"
      }
    ]
  }
ben@glitnir 15:19:49 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler queue
┌─────────────────────────────────────────┬───────────┬────────┬──────┬─────────┬──────┬───────┬──────────────────┐
│ JOBID                                   │ PARTITION │ NAME   │ USER │ ST      │ TIME │ NODES │ NODELIST(REASON) │
├─────────────────────────────────────────┼───────────┼────────┼──────┼─────────┼──────┼───────┼──────────────────┤
│ 19506270-22ca-4199-9204-6d43222319c5[1] │ all       │ (none) │ TBD  │ RUNNING │ TBD  │ 1     │ node01           │
│ 19506270-22ca-4199-9204-6d43222319c5[2] │ all       │ (none) │ TBD  │ RUNNING │ TBD  │ 1     │ node02           │
└─────────────────────────────────────────┴───────────┴────────┴──────┴─────────┴──────┴───────┴──────────────────┘
ben@glitnir 15:19:51 flight-scheduler (dev/remove-tasks *$%=)
```

This time nothing to do with job steps.

benarmston commented:
When running the same job script as a non-array job cancellation works fine.

```
ben@glitnir 15:24:23 flight-scheduler (dev/remove-tasks *$%=)
$ cat tmp/super-simple-array.sh 
#!/bin/bash

#SBATCH -N 1

echo "sleeping"
sleep 120
echo "done sleeping"
ben@glitnir 15:24:27 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler batch --nodes 1 tmp/super-simple-array.sh
Submitted batch job dedc667d-e817-4c84-9ff8-88c460a18e6b
ben@glitnir 15:24:35 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler queue
┌──────────────────────────────────────┬───────────┬───────────────────────┬──────┬─────────┬──────┬───────┬──────────────────┐
│ JOBID                                │ PARTITION │ NAME                  │ USER │ ST      │ TIME │ NODES │ NODELIST(REASON) │
├──────────────────────────────────────┼───────────┼───────────────────────┼──────┼─────────┼──────┼───────┼──────────────────┤
│ dedc667d-e817-4c84-9ff8-88c460a18e6b │ all       │ super-simple-array.sh │ TBD  │ RUNNING │ TBD  │ 1     │ node01           │
└──────────────────────────────────────┴───────────┴───────────────────────┴──────┴─────────┴──────┴───────┴──────────────────┘
ben@glitnir 15:24:39 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler cancel dedc667d-e817-4c84-9ff8-88c460a18e6b
ben@glitnir 15:24:46 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler queue

ben@glitnir 15:24:48 flight-scheduler (dev/remove-tasks *$%=)
```

benarmston commented:
If I happen to know the array task's ID I can attempt to cancel it with that ID.  I get different bad behaviour when doing so.

```
ben@glitnir 15:26:32 flight-scheduler (dev/remove-tasks *$%=)
$ cat tmp/super-simple-array.sh 
#!/bin/bash

#SBATCH -N 1

echo "sleeping"
sleep 120
echo "done sleeping"
ben@glitnir 15:26:39 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler batch --array 1,2  --nodes 1 tmp/super-simple-array.sh
Submitted batch job d03cf7fe-5db5-4ccf-b6cd-6fb60a00b83d[-2]
ben@glitnir 15:26:46 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler queue
┌─────────────────────────────────────────┬───────────┬────────┬──────┬─────────┬──────┬───────┬──────────────────┐
│ JOBID                                   │ PARTITION │ NAME   │ USER │ ST      │ TIME │ NODES │ NODELIST(REASON) │
├─────────────────────────────────────────┼───────────┼────────┼──────┼─────────┼──────┼───────┼──────────────────┤
│ d03cf7fe-5db5-4ccf-b6cd-6fb60a00b83d[1] │ all       │ (none) │ TBD  │ RUNNING │ TBD  │ 1     │ node01           │
│ d03cf7fe-5db5-4ccf-b6cd-6fb60a00b83d[2] │ all       │ (none) │ TBD  │ RUNNING │ TBD  │ 1     │ node02           │
└─────────────────────────────────────────┴───────────┴────────┴──────┴─────────┴──────┴───────┴──────────────────┘
ben@glitnir 15:26:48 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler cancel 4945b663-c735-451d-a4df-6c5d7ec7afa1
ben@glitnir 15:27:01 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler queue
┌─────────────────────────────────────────┬───────────┬────────┬──────┬─────────┬──────┬───────┬──────────────────┐
│ JOBID                                   │ PARTITION │ NAME   │ USER │ ST      │ TIME │ NODES │ NODELIST(REASON) │
├─────────────────────────────────────────┼───────────┼────────┼──────┼─────────┼──────┼───────┼──────────────────┤
│ d03cf7fe-5db5-4ccf-b6cd-6fb60a00b83d[1] │ all       │ (none) │ TBD  │ RUNNING │ TBD  │ 1     │ node01           │
│ d03cf7fe-5db5-4ccf-b6cd-6fb60a00b83d[2] │ all       │ (none) │ TBD  │ RUNNING │ TBD  │ 1     │ node02           │
└─────────────────────────────────────────┴───────────┴────────┴──────┴─────────┴──────┴───────┴──────────────────┘
ben@glitnir 15:27:03 flight-scheduler (dev/remove-tasks *$%=)
```

So the job is found this time, but it isn't cancelled.

benarmston commented:
It looks like the controller is a bit confused about whether it cancels `ARRAY_JOB`s or `ARRAY_TASK`s.

benarmston commented:
Cancelling pending array tasks doesn't quite work either:

```
ben@glitnir 15:35:13 flight-scheduler (dev/remove-tasks *$%=)
$ cat tmp/super-simple-array.sh 
#!/bin/bash

#SBATCH -N 1

echo "sleeping"
sleep 120
echo "done sleeping"
ben@glitnir 15:35:25 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler batch --array 1,2  --nodes 1 tmp/super-simple-array.sh
Submitted batch job eea13a47-58a8-44df-8f05-dc14db48e68e[2]
ben@glitnir 15:35:31 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler queue
┌─────────────────────────────────────────┬───────────┬───────────────────────┬──────┬─────────┬──────┬───────┬──────────────────┐
│ JOBID                                   │ PARTITION │ NAME                  │ USER │ ST      │ TIME │ NODES │ NODELIST(REASON) │
├─────────────────────────────────────────┼───────────┼───────────────────────┼──────┼─────────┼──────┼───────┼──────────────────┤
│ eea13a47-58a8-44df-8f05-dc14db48e68e[2] │ all       │ super-simple-array.sh │ TBD  │ PENDING │ TBD  │ 1     │ (Resources)      │
│ eea13a47-58a8-44df-8f05-dc14db48e68e[1] │ all       │ (none)                │ TBD  │ RUNNING │ TBD  │ 1     │ node01           │
└─────────────────────────────────────────┴───────────┴───────────────────────┴──────┴─────────┴──────┴───────┴──────────────────┘
ben@glitnir 15:35:37 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler cancel eea13a47-58a8-44df-8f05-dc14db48e68e
ben@glitnir 15:35:45 flight-scheduler (dev/remove-tasks *$%=)
$ ./bin/scheduler queue

ben@glitnir 15:35:50 flight-scheduler (dev/remove-tasks *$%=)
```

The above output is exactly what I'd expect, however, looking at the daemon logs we have:

```
16m25s     info: Processing message {:command=>"JOB_ALLOCATED", :environment=>{:CLUSTER_NAME=>"", :JOB_ID=>"fab82d9b-7946-4169-b4eb-76189b5f63d9", :JOB_PARTITION=>"all", :JOB_NODES=>"1", :JOB_NODELIST=>"node01", :NODENAME=>"node01", :ARRAY_JOB_ID=>"eea13a47-58a8-44df-8f05-dc14db48e68e", :ARRAY_TASK_ID=>"1", :ARRAY_TASK_COUNT=>"2", :ARRAY_TASK_MAX=>"2", :ARRAY_TASK_MIN=>"1"}, :job_id=>"fab82d9b-7946-4169-b4eb-76189b5f63d9", :username=>"ben"} [pid=32245] [2020-10-08 15:35:31 +0100]
16m25s    debug: Environment: CLUSTER_NAME=
JOB_ID=fab82d9b-7946-4169-b4eb-76189b5f63d9
JOB_PARTITION=all
JOB_NODES=1
JOB_NODELIST=node01
NODENAME=node01
ARRAY_JOB_ID=eea13a47-58a8-44df-8f05-dc14db48e68e
ARRAY_TASK_ID=1
ARRAY_TASK_COUNT=2
ARRAY_TASK_MAX=2
ARRAY_TASK_MIN=1 [pid=32245] [2020-10-08 15:35:31 +0100]
16m25s    debug: Processed message {:command=>"JOB_ALLOCATED", :environment=>{:CLUSTER_NAME=>"", :JOB_ID=>"fab82d9b-7946-4169-b4eb-76189b5f63d9", :JOB_PARTITION=>"all", :JOB_NODES=>"1", :JOB_NODELIST=>"node01", :NODENAME=>"node01", :ARRAY_JOB_ID=>"eea13a47-58a8-44df-8f05-dc14db48e68e", :ARRAY_TASK_ID=>"1", :ARRAY_TASK_COUNT=>"2", :ARRAY_TASK_MAX=>"2", :ARRAY_TASK_MIN=>"1"}, :job_id=>"fab82d9b-7946-4169-b4eb-76189b5f63d9", :username=>"ben"} [pid=32245] [2020-10-08 15:35:31 +0100]
16m25s     info: Processing message {:command=>"RUN_SCRIPT", :arguments=>[], :array_job_id=>"eea13a47-58a8-44df-8f05-dc14db48e68e", :array_task_id=>"fab82d9b-7946-4169-b4eb-76189b5f63d9", :job_id=>"fab82d9b-7946-4169-b4eb-76189b5f63d9", :script=>"#!/bin/bash\n\n#SBATCH -N 1\n\necho \"sleeping\"\nsleep 120\necho \"done sleeping\"\n", :stderr_path=>"flight-scheduler-eea13a47-58a8-44df-8f05-dc14db48e68e_1.out", :stdout_path=>"flight-scheduler-eea13a47-58a8-44df-8f05-dc14db48e68e_1.out"} [pid=32245] [2020-10-08 15:35:31 +0100]
16m25s    debug: Running script for job:fab82d9b-7946-4169-b4eb-76189b5f63d9 script:#!/bin/bash

#SBATCH -N 1

echo "sleeping"
sleep 120
echo "done sleeping"
 arguments:[] [pid=32245] [2020-10-08 15:35:31 +0100]
16m25s    debug: Processed message {:command=>"RUN_SCRIPT", :arguments=>[], :array_job_id=>"eea13a47-58a8-44df-8f05-dc14db48e68e", :array_task_id=>"fab82d9b-7946-4169-b4eb-76189b5f63d9", :job_id=>"fab82d9b-7946-4169-b4eb-76189b5f63d9", :script=>"#!/bin/bash\n\n#SBATCH -N 1\n\necho \"sleeping\"\nsleep 120\necho \"done sleeping\"\n", :stderr_path=>"flight-scheduler-eea13a47-58a8-44df-8f05-dc14db48e68e_1.out", :stdout_path=>"flight-scheduler-eea13a47-58a8-44df-8f05-dc14db48e68e_1.out"} [pid=32245] [2020-10-08 15:35:31 +0100]

<Cancellation request is sent here>

16m39s     info: Processing message {:command=>"JOB_CANCELLED", :job_id=>"fab82d9b-7946-4169-b4eb-76189b5f63d9"} [pid=32245] [2020-10-08 15:35:45 +0100]
16m39s     info: Cancelling job:fab82d9b-7946-4169-b4eb-76189b5f63d9 [pid=32245] [2020-10-08 15:35:45 +0100]
16m39s    debug: Processed message {:command=>"JOB_CANCELLED", :job_id=>"fab82d9b-7946-4169-b4eb-76189b5f63d9"} [pid=32245] [2020-10-08 15:35:45 +0100]
16m39s     info: Completed job fab82d9b-7946-4169-b4eb-76189b5f63d9 [pid=32245] [2020-10-08 15:35:45 +0100]
16m39s     info: Processing message {:command=>"JOB_DEALLOCATED", :job_id=>"fab82d9b-7946-4169-b4eb-76189b5f63d9"} [pid=32245] [2020-10-08 15:35:45 +0100]
16m39s     info: Deallocating job:fab82d9b-7946-4169-b4eb-76189b5f63d9 [pid=32245] [2020-10-08 15:35:45 +0100]
16m39s    debug: Processed message {:command=>"JOB_DEALLOCATED", :job_id=>"fab82d9b-7946-4169-b4eb-76189b5f63d9"} [pid=32245] [2020-10-08 15:35:45 +0100]
16m39s     info: Processing message {:command=>"JOB_ALLOCATED", :environment=>{:CLUSTER_NAME=>"", :JOB_ID=>"029b67c4-e4b6-42fe-9531-0b48e112ef98", :JOB_PARTITION=>"all", :JOB_NODES=>"0", :JOB_NODELIST=>"", :NODENAME=>"node01", :ARRAY_JOB_ID=>"eea13a47-58a8-44df-8f05-dc14db48e68e", :ARRAY_TASK_ID=>"2", :ARRAY_TASK_COUNT=>"2", :ARRAY_TASK_MAX=>"2", :ARRAY_TASK_MIN=>"1"}, :job_id=>"029b67c4-e4b6-42fe-9531-0b48e112ef98", :username=>"ben"} [pid=32245] [2020-10-08 15:35:45 +0100]
16m39s    debug: Environment: CLUSTER_NAME=
JOB_ID=029b67c4-e4b6-42fe-9531-0b48e112ef98
JOB_PARTITION=all
JOB_NODES=0
JOB_NODELIST=
NODENAME=node01
ARRAY_JOB_ID=eea13a47-58a8-44df-8f05-dc14db48e68e
ARRAY_TASK_ID=2
ARRAY_TASK_COUNT=2
ARRAY_TASK_MAX=2
ARRAY_TASK_MIN=1 [pid=32245] [2020-10-08 15:35:45 +0100]
16m39s    debug: Processed message {:command=>"JOB_ALLOCATED", :environment=>{:CLUSTER_NAME=>"", :JOB_ID=>"029b67c4-e4b6-42fe-9531-0b48e112ef98", :JOB_PARTITION=>"all", :JOB_NODES=>"0", :JOB_NODELIST=>"", :NODENAME=>"node01", :ARRAY_JOB_ID=>"eea13a47-58a8-44df-8f05-dc14db48e68e", :ARRAY_TASK_ID=>"2", :ARRAY_TASK_COUNT=>"2", :ARRAY_TASK_MAX=>"2", :ARRAY_TASK_MIN=>"1"}, :job_id=>"029b67c4-e4b6-42fe-9531-0b48e112ef98", :username=>"ben"} [pid=32245] [2020-10-08 15:35:45 +0100]
16m39s     info: Processing message {:command=>"RUN_SCRIPT", :arguments=>[], :array_job_id=>"eea13a47-58a8-44df-8f05-dc14db48e68e", :array_task_id=>"029b67c4-e4b6-42fe-9531-0b48e112ef98", :job_id=>"029b67c4-e4b6-42fe-9531-0b48e112ef98", :script=>nil, :stderr_path=>"flight-scheduler-eea13a47-58a8-44df-8f05-dc14db48e68e_2.out", :stdout_path=>"flight-scheduler-eea13a47-58a8-44df-8f05-dc14db48e68e_2.out"} [pid=32245] [2020-10-08 15:35:45 +0100]
16m39s    debug: Running script for job:029b67c4-e4b6-42fe-9531-0b48e112ef98 script: arguments:[] [pid=32245] [2020-10-08 15:35:45 +0100]
16m39s     warn: Error running script job:029b67c4-e4b6-42fe-9531-0b48e112ef98 An unexpected error has occurred! The batch script does not appear
to be in a valid state. [pid=32245] [2020-10-08 15:35:45 +0100]
16m39s    debug: Processed message {:command=>"RUN_SCRIPT", :arguments=>[], :array_job_id=>"eea13a47-58a8-44df-8f05-dc14db48e68e", :array_task_id=>"029b67c4-e4b6-42fe-9531-0b48e112ef98", :job_id=>"029b67c4-e4b6-42fe-9531-0b48e112ef98", :script=>nil, :stderr_path=>"flight-scheduler-eea13a47-58a8-44df-8f05-dc14db48e68e_2.out", :stdout_path=>"flight-scheduler-eea13a47-58a8-44df-8f05-dc14db48e68e_2.out"} [pid=32245] [2020-10-08 15:35:46 +0100]
16m39s     info: Processing message {:command=>"JOB_DEALLOCATED", :job_id=>"029b67c4-e4b6-42fe-9531-0b48e112ef98"} [pid=32245] [2020-10-08 15:35:46 +0100]
16m39s     info: Deallocating job:029b67c4-e4b6-42fe-9531-0b48e112ef98 [pid=32245] [2020-10-08 15:35:46 +0100]
16m39s    debug: Processed message {:command=>"JOB_DEALLOCATED", :job_id=>"029b67c4-e4b6-42fe-9531-0b48e112ef98"} [pid=32245] [2020-10-08 15:35:46 +0100]
```

The controller logs are:

```
16m27s    debug: Falcon::Server [oid=0x1068] [pid=32213] [2020-10-08 15:35:31 +0100]
               | Incoming connnection from #<Addrinfo: 127.0.0.1:35230 TCP> to Async::HTTP::Protocol::HTTP1
16m27s    debug: Added job eea13a47-58a8-44df-8f05-dc14db48e68e to FifoScheduler [pid=32213] [2020-10-08 15:35:31 +0100]
16m27s     info: Attempting to allocate rescources to jobs [pid=32213] [2020-10-08 15:35:31 +0100]
16m27s    debug: Queued jobs ["eea13a47-58a8-44df-8f05-dc14db48e68e"] [pid=32213] [2020-10-08 15:35:31 +0100]
16m27s    debug: Allocated jobs [] [pid=32213] [2020-10-08 15:35:31 +0100]
16m27s    debug: Connected nodes ["node01"] [pid=32213] [2020-10-08 15:35:31 +0100]
16m27s    debug: Allocated nodes [] [pid=32213] [2020-10-08 15:35:31 +0100]
16m27s     info: Allocated node01 to job fab82d9b-7946-4169-b4eb-76189b5f63d9 [pid=32213] [2020-10-08 15:35:31 +0100]
16m27s    debug: Initializing array task 1 for eea13a47-58a8-44df-8f05-dc14db48e68e on node01 [pid=32213] [2020-10-08 15:35:31 +0100]
16m27s    debug: Initialized array task 1 for eea13a47-58a8-44df-8f05-dc14db48e68e on node01 [pid=32213] [2020-10-08 15:35:31 +0100]
16m27s    debug: Sending batch script for array task 1 for eea13a47-58a8-44df-8f05-dc14db48e68e to node01 [pid=32213] [2020-10-08 15:35:31 +0100]
16m27s    debug: Sent batch script for array task 1 for eea13a47-58a8-44df-8f05-dc14db48e68e to node01 [pid=32213] [2020-10-08 15:35:31 +0100]
16m27s     info: 127.0.0.1 - - [08/Oct/2020:15:35:31 +0100] "POST /v0/jobs HTTP/1.1" 201 892 0.0017
 [pid=32213] [2020-10-08 15:35:31 +0100]
16m27s    debug: Async::HTTP::Protocol::HTTP1::Server [oid=0x15b8] [pid=32213] [2020-10-08 15:35:31 +0100]
               | Closing connection
16m33s    debug: Falcon::Server [oid=0x1068] [pid=32213] [2020-10-08 15:35:37 +0100]
               | Incoming connnection from #<Addrinfo: 127.0.0.1:35232 TCP> to Async::HTTP::Protocol::HTTP1
16m33s     info: 127.0.0.1 - - [08/Oct/2020:15:35:37 +0100] "GET /v0/jobs?include=partition%2Callocated-nodes HTTP/1.1" 200 2917 0.0010
 [pid=32213] [2020-10-08 15:35:37 +0100]
16m33s    debug: Async::HTTP::Protocol::HTTP1::Server [oid=0x15e0] [pid=32213] [2020-10-08 15:35:37 +0100]
               | Closing connection
16m40s    debug: Falcon::Server [oid=0x1068] [pid=32213] [2020-10-08 15:35:45 +0100]
               | Incoming connnection from #<Addrinfo: 127.0.0.1:35236 TCP> to Async::HTTP::Protocol::HTTP1
16m40s     info: Cancelling running array job eea13a47-58a8-44df-8f05-dc14db48e68e [pid=32213] [2020-10-08 15:35:45 +0100]
16m40s    debug: Job cancellation for task 1 of job eea13a47-58a8-44df-8f05-dc14db48e68e sent to node01 [pid=32213] [2020-10-08 15:35:45 +0100]
16m40s     info: 127.0.0.1 - - [08/Oct/2020:15:35:45 +0100] "DELETE /v0/jobs/eea13a47-58a8-44df-8f05-dc14db48e68e HTTP/1.1" 204 - 0.0007
 [pid=32213] [2020-10-08 15:35:45 +0100]
16m40s    debug: Async::HTTP::Protocol::HTTP1::Server [oid=0x1608] [pid=32213] [2020-10-08 15:35:45 +0100]
               | Closing connection
16m41s     info: Processing message {:command=>"NODE_FAILED_ARRAY_TASK", :array_job_id=>"eea13a47-58a8-44df-8f05-dc14db48e68e", :array_task_id=>"fab82d9b-7946-4169-b4eb-76189b5f63d9"} [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s     info: Node node01 failed task 1 for job eea13a47-58a8-44df-8f05-dc14db48e68e [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s    debug: Job deallocation for fab82d9b-7946-4169-b4eb-76189b5f63d9 sent to node01 [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s     info: Processing message {:command=>"NODE_DEALLOCATED", :job_id=>"fab82d9b-7946-4169-b4eb-76189b5f63d9"} [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s    debug: Removed job fab82d9b-7946-4169-b4eb-76189b5f63d9 from FifoScheduler [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s     info: Attempting to allocate rescources to jobs [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s    debug: Queued jobs ["eea13a47-58a8-44df-8f05-dc14db48e68e"] [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s    debug: Allocated jobs [] [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s    debug: Connected nodes ["node01"] [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s    debug: Allocated nodes [] [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s     info: Allocated node01 to job 029b67c4-e4b6-42fe-9531-0b48e112ef98 [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s    debug: Initializing array task 2 for eea13a47-58a8-44df-8f05-dc14db48e68e on node01 [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s    debug: Initialized array task 2 for eea13a47-58a8-44df-8f05-dc14db48e68e on node01 [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s    debug: Sending batch script for array task 2 for eea13a47-58a8-44df-8f05-dc14db48e68e to node01 [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s    debug: Sent batch script for array task 2 for eea13a47-58a8-44df-8f05-dc14db48e68e to node01 [pid=32213] [2020-10-08 15:35:45 +0100]
16m41s     info: Processing message {:command=>"NODE_FAILED_ARRAY_TASK", :array_job_id=>"eea13a47-58a8-44df-8f05-dc14db48e68e", :array_task_id=>"029b67c4-e4b6-42fe-9531-0b48e112ef98"} [pid=32213] [2020-10-08 15:35:46 +0100]
16m41s     info: Node node01 failed task 2 for job eea13a47-58a8-44df-8f05-dc14db48e68e [pid=32213] [2020-10-08 15:35:46 +0100]
16m41s    debug: Job deallocation for 029b67c4-e4b6-42fe-9531-0b48e112ef98 sent to node01 [pid=32213] [2020-10-08 15:35:46 +0100]
16m41s     info: Processing message {:command=>"NODE_DEALLOCATED", :job_id=>"029b67c4-e4b6-42fe-9531-0b48e112ef98"} [pid=32213] [2020-10-08 15:35:46 +0100]
16m41s    debug: Removed job 029b67c4-e4b6-42fe-9531-0b48e112ef98 from FifoScheduler [pid=32213] [2020-10-08 15:35:46 +0100]
16m41s     info: Attempting to allocate rescources to jobs [pid=32213] [2020-10-08 15:35:46 +0100]
16m41s    debug: Queued jobs [] [pid=32213] [2020-10-08 15:35:46 +0100]
16m41s    debug: Allocated jobs [] [pid=32213] [2020-10-08 15:35:46 +0100]
16m41s    debug: Connected nodes ["node01"] [pid=32213] [2020-10-08 15:35:46 +0100]
16m41s    debug: Allocated nodes [] [pid=32213] [2020-10-08 15:35:46 +0100]
16m41s     info: Processing message {:command=>"NODE_DEALLOCATED", :job_id=>"fab82d9b-7946-4169-b4eb-76189b5f63d9"} [pid=32213] [2020-10-08 15:35:46 +0100]
16m46s    debug: Falcon::Server [oid=0x1068] [pid=32213] [2020-10-08 15:35:50 +0100]
               | Incoming connnection from #<Addrinfo: 127.0.0.1:35238 TCP> to Async::HTTP::Protocol::HTTP1
16m46s     info: 127.0.0.1 - - [08/Oct/2020:15:35:50 +0100] "GET /v0/jobs?include=partition%2Callocated-nodes HTTP/1.1" 200 83 0.0007
 [pid=32213] [2020-10-08 15:35:50 +0100]
16m46s    debug: Async::HTTP::Protocol::HTTP1::Server [oid=0x1644] [pid=32213] [2020-10-08 15:35:50 +0100]
               | Closing connection

```

It looks like the controller is attempting to run the the second `ARRAY_TASK` for the `ARRAY_JOB`.  We have a `CANCELLING` state for jobs that can be used to cover this case.  Also the scheduler no longer appears to limit itself to pending jobs.

WilliamMcCumstie committed:
Implicitly always find the array_job

Finding array tasks is currently not supported

WilliamMcCumstie committed:
Prevent the scheduler starting cancelled jobs

The scheduler relies on the enumerator to start jobs, thus by
setting it to return nils it can't start anymore

benarmston commented:
Cancelling pending jobs is somewhat broken.  The jobs are never removed from the scheduler.

Here I add a job requesting more nodes than are available.  It will remain in the pending state.  When I cancel it it remains in the queue.  Fortunately, it doesn't block any other jobs from being scheduled.

```
$ cat tmp/super-simple.sh 
#!/bin/bash

#SBATCH -N 1

echo "sleeping"
sleep 120
echo "done sleeping"
ben@glitnir 10:29:17 flight-scheduler (dev/remove-tasks *+$%=)
$ ./bin/scheduler batch  --nodes 2 tmp/super-simple.sh ; ./bin/scheduler queue
Submitted batch job 3d81dcba-a25b-4f76-aac2-4c147186c165
┌──────────────────────────────────────┬───────────┬─────────────────┬──────┬─────────┬──────┬───────┬──────────────────┐
│ JOBID                                │ PARTITION │ NAME            │ USER │ ST      │ TIME │ NODES │ NODELIST(REASON) │
├──────────────────────────────────────┼───────────┼─────────────────┼──────┼─────────┼──────┼───────┼──────────────────┤
│ 3d81dcba-a25b-4f76-aac2-4c147186c165 │ all       │ super-simple.sh │ TBD  │ PENDING │ TBD  │ 2     │ (Resources)      │
└──────────────────────────────────────┴───────────┴─────────────────┴──────┴─────────┴──────┴───────┴──────────────────┘
ben@glitnir 10:29:25 flight-scheduler (dev/remove-tasks *+$%=)
$ ./bin/scheduler cancel 3d81dcba-a25b-4f76-aac2-4c147186c165
ben@glitnir 10:29:36 flight-scheduler (dev/remove-tasks *+$%=)
$  ./bin/scheduler queue
┌──────────────────────────────────────┬───────────┬─────────────────┬──────┬───────────┬──────┬───────┬──────────────────┐
│ JOBID                                │ PARTITION │ NAME            │ USER │ ST        │ TIME │ NODES │ NODELIST(REASON) │
├──────────────────────────────────────┼───────────┼─────────────────┼──────┼───────────┼──────┼───────┼──────────────────┤
│ 3d81dcba-a25b-4f76-aac2-4c147186c165 │ all       │ super-simple.sh │ TBD  │ CANCELLED │ TBD  │ 2     │ (none)           │
└──────────────────────────────────────┴───────────┴─────────────────┴──────┴───────────┴──────┴───────┴──────────────────┘
```

Here I add a job which will run (the above job is still stuck in the queue).  Whilst it is running, I add a second job.  The second job is pending on job priority.  Whilst it is pending I cancel it and it remains in the queue.

```
ben@glitnir 10:30:06 flight-scheduler (dev/remove-tasks *+$%=)
$ ./bin/scheduler batch  --nodes 1 tmp/super-simple.sh ; ./bin/scheduler queue
Submitted batch job 78ab5a42-e2cd-421b-9bee-245d1f35c8ca
┌──────────────────────────────────────┬───────────┬─────────────────┬──────┬───────────┬──────┬───────┬──────────────────┐
│ JOBID                                │ PARTITION │ NAME            │ USER │ ST        │ TIME │ NODES │ NODELIST(REASON) │
├──────────────────────────────────────┼───────────┼─────────────────┼──────┼───────────┼──────┼───────┼──────────────────┤
│ 3d81dcba-a25b-4f76-aac2-4c147186c165 │ all       │ super-simple.sh │ TBD  │ CANCELLED │ TBD  │ 2     │ (none)           │
│ 78ab5a42-e2cd-421b-9bee-245d1f35c8ca │ all       │ super-simple.sh │ TBD  │ RUNNING   │ TBD  │ 1     │ node01           │
└──────────────────────────────────────┴───────────┴─────────────────┴──────┴───────────┴──────┴───────┴──────────────────┘
ben@glitnir 10:30:17 flight-scheduler (dev/remove-tasks *+$%=)
$ ./bin/scheduler batch  --nodes 1 tmp/super-simple.sh ; ./bin/scheduler queue
Submitted batch job 03449535-678d-49fb-abc3-eb923ff78de6
┌──────────────────────────────────────┬───────────┬─────────────────┬──────┬───────────┬──────┬───────┬──────────────────┐
│ JOBID                                │ PARTITION │ NAME            │ USER │ ST        │ TIME │ NODES │ NODELIST(REASON) │
├──────────────────────────────────────┼───────────┼─────────────────┼──────┼───────────┼──────┼───────┼──────────────────┤
│ 3d81dcba-a25b-4f76-aac2-4c147186c165 │ all       │ super-simple.sh │ TBD  │ CANCELLED │ TBD  │ 2     │ (none)           │
│ 78ab5a42-e2cd-421b-9bee-245d1f35c8ca │ all       │ super-simple.sh │ TBD  │ RUNNING   │ TBD  │ 1     │ node01           │
│ 03449535-678d-49fb-abc3-eb923ff78de6 │ all       │ super-simple.sh │ TBD  │ PENDING   │ TBD  │ 1     │ (Resources)      │
└──────────────────────────────────────┴───────────┴─────────────────┴──────┴───────────┴──────┴───────┴──────────────────┘
ben@glitnir 10:30:22 flight-scheduler (dev/remove-tasks *+$%=)
$  ./bin/scheduler cancel 03449535-678d-49fb-abc3-eb923ff78de6
ben@glitnir 10:30:30 flight-scheduler (dev/remove-tasks *+$%=)
$  ./bin/scheduler queue
┌──────────────────────────────────────┬───────────┬─────────────────┬──────┬───────────┬──────┬───────┬──────────────────┐
│ JOBID                                │ PARTITION │ NAME            │ USER │ ST        │ TIME │ NODES │ NODELIST(REASON) │
├──────────────────────────────────────┼───────────┼─────────────────┼──────┼───────────┼──────┼───────┼──────────────────┤
│ 3d81dcba-a25b-4f76-aac2-4c147186c165 │ all       │ super-simple.sh │ TBD  │ CANCELLED │ TBD  │ 2     │ (none)           │
│ 78ab5a42-e2cd-421b-9bee-245d1f35c8ca │ all       │ super-simple.sh │ TBD  │ RUNNING   │ TBD  │ 1     │ node01           │
│ 03449535-678d-49fb-abc3-eb923ff78de6 │ all       │ super-simple.sh │ TBD  │ CANCELLED │ TBD  │ 1     │ (none)           │
└──────────────────────────────────────┴───────────┴─────────────────┴──────┴───────────┴──────┴───────┴──────────────────┘
```

Here I add a job which will run.  Whilst it is running, I cancel it.  Everything works as it should.

```
ben@glitnir 10:33:48 flight-scheduler (dev/remove-tasks *+$%=)
$  ./bin/scheduler queue
┌──────────────────────────────────────┬───────────┬─────────────────┬──────┬───────────┬──────┬───────┬──────────────────┐
│ JOBID                                │ PARTITION │ NAME            │ USER │ ST        │ TIME │ NODES │ NODELIST(REASON) │
├──────────────────────────────────────┼───────────┼─────────────────┼──────┼───────────┼──────┼───────┼──────────────────┤
│ 3d81dcba-a25b-4f76-aac2-4c147186c165 │ all       │ super-simple.sh │ TBD  │ CANCELLED │ TBD  │ 2     │ (none)           │
│ 03449535-678d-49fb-abc3-eb923ff78de6 │ all       │ super-simple.sh │ TBD  │ CANCELLED │ TBD  │ 1     │ (none)           │
└──────────────────────────────────────┴───────────┴─────────────────┴──────┴───────────┴──────┴───────┴──────────────────┘
ben@glitnir 10:33:51 flight-scheduler (dev/remove-tasks *+$%=)
$ ./bin/scheduler batch  --nodes 1 tmp/super-simple.sh ; ./bin/scheduler queue
Submitted batch job 762ca0d9-8ac1-4cad-a126-4a6be4f90ffa
┌──────────────────────────────────────┬───────────┬─────────────────┬──────┬───────────┬──────┬───────┬──────────────────┐
│ JOBID                                │ PARTITION │ NAME            │ USER │ ST        │ TIME │ NODES │ NODELIST(REASON) │
├──────────────────────────────────────┼───────────┼─────────────────┼──────┼───────────┼──────┼───────┼──────────────────┤
│ 3d81dcba-a25b-4f76-aac2-4c147186c165 │ all       │ super-simple.sh │ TBD  │ CANCELLED │ TBD  │ 2     │ (none)           │
│ 03449535-678d-49fb-abc3-eb923ff78de6 │ all       │ super-simple.sh │ TBD  │ CANCELLED │ TBD  │ 1     │ (none)           │
│ 762ca0d9-8ac1-4cad-a126-4a6be4f90ffa │ all       │ super-simple.sh │ TBD  │ RUNNING   │ TBD  │ 1     │ node01           │
└──────────────────────────────────────┴───────────┴─────────────────┴──────┴───────────┴──────┴───────┴──────────────────┘
ben@glitnir 10:33:57 flight-scheduler (dev/remove-tasks *+$%=)
$  ./bin/scheduler cancel 762ca0d9-8ac1-4cad-a126-4a6be4f90ffa
ben@glitnir 10:34:02 flight-scheduler (dev/remove-tasks *+$%=)
$  ./bin/scheduler queue
┌──────────────────────────────────────┬───────────┬─────────────────┬──────┬───────────┬──────┬───────┬──────────────────┐
│ JOBID                                │ PARTITION │ NAME            │ USER │ ST        │ TIME │ NODES │ NODELIST(REASON) │
├──────────────────────────────────────┼───────────┼─────────────────┼──────┼───────────┼──────┼───────┼──────────────────┤
│ 3d81dcba-a25b-4f76-aac2-4c147186c165 │ all       │ super-simple.sh │ TBD  │ CANCELLED │ TBD  │ 2     │ (none)           │
│ 03449535-678d-49fb-abc3-eb923ff78de6 │ all       │ super-simple.sh │ TBD  │ CANCELLED │ TBD  │ 1     │ (none)           │
└──────────────────────────────────────┴───────────┴─────────────────┴──────┴───────────┴──────┴───────┴──────────────────┘
```

---

It is a similar situation for array jobs.

Here I cancel an `ARRAY_JOB` whilst it is pending.  The `ARRAY_JOB` gets cancelled, but is still stuck in the queue.  It looks like one of its `ARRAY_TASKS` has disappeared.

```
ben@glitnir 10:35:29 flight-scheduler (dev/remove-tasks *+$%=)
$ ./bin/scheduler batch --array 1,2 --nodes 2 tmp/super-simple.sh ; ./bin/scheduler queue
Submitted batch job d639ba1b-f1db-4c02-b1df-274d04391adc[1-2]
┌───────────────────────────────────────────┬───────────┬─────────────────┬──────┬─────────┬──────┬───────┬──────────────────┐
│ JOBID                                     │ PARTITION │ NAME            │ USER │ ST      │ TIME │ NODES │ NODELIST(REASON) │
├───────────────────────────────────────────┼───────────┼─────────────────┼──────┼─────────┼──────┼───────┼──────────────────┤
│ d639ba1b-f1db-4c02-b1df-274d04391adc[1-2] │ all       │ super-simple.sh │ TBD  │ PENDING │ TBD  │ 2     │ (Resources)      │
└───────────────────────────────────────────┴───────────┴─────────────────┴──────┴─────────┴──────┴───────┴──────────────────┘
ben@glitnir 10:35:52 flight-scheduler (dev/remove-tasks *+$%=)
$ ./bin/scheduler cancel d639ba1b-f1db-4c02-b1df-274d04391adc
ben@glitnir 10:36:01 flight-scheduler (dev/remove-tasks *+$%=)
$  ./bin/scheduler queue
┌──────────────────────────────────────────┬───────────┬─────────────────┬──────┬───────────┬──────┬───────┬──────────────────┐
│ JOBID                                    │ PARTITION │ NAME            │ USER │ ST        │ TIME │ NODES │ NODELIST(REASON) │
├──────────────────────────────────────────┼───────────┼─────────────────┼──────┼───────────┼──────┼───────┼──────────────────┤
│ d639ba1b-f1db-4c02-b1df-274d04391adc[-2] │ all       │ super-simple.sh │ TBD  │ CANCELLED │ TBD  │ 2     │ (none)           │
└──────────────────────────────────────────┴───────────┴─────────────────┴──────┴───────────┴──────┴───────┴──────────────────┘
```

WilliamMcCumstie committed:
Remember to remove pending jobs from the schduler

benarmston commented:
It looks like this is now working.

There are parts of this PR that I like and other parts that I think are not quite right.

The removal of `Task` seems like a step in the right direction.  That `node_{completed,failed}_{job,task}` are starting to look very similar, suggests that the controller<->daemon communication is getting close to where it should be.  As does the similarity of `Submission::BatchJob` and `Submission::ArrayTask`.

I think that the difficulty in getting this to work correctly is an indication that something isn't quite right.  My feeling is that it is the use of `Enumerator` and the attempt to decouple the scheduler from job types.

Those decisions have mean that there is no longer a single source of truth for a job's state.  For example, if we want to know if a job is being cancelled, we cannot check `job.state == 'CANCELLING'`.  I tried to write down what would need to do to determine that, but ended up in a tangled mess including, the `Enumerator` returned by the job's `to_enum` method, the `Enumerator` held by the scheduler for that job, which may or may not be the same, and enumerator's for `ARRAY_TASK`s which always return `nil`.

I understand why you'd want to decouple the scheduler from job types, but losing the single source of truth has resulted in an overal gain in code complexity.

I also think that correct scheduling of array jobs is a core part of the schedulers functionality.  Here I'm thinking of things like limiting the number of concurrently running array tasks for a given array job; prefering to schedule non-exclusive array tasks on the same node rather than on different nodes, and many others.  I appreciate these are features that we don't have at the moment.  But they suggest to me that scheduling array tasks is a special case that should be handled by the scheduler, not something that should be hidden from the scheduler through some abstraction.

My other issues with `Enumerator`s are that I believe they are not serializable/deserializable.  I suspect that this will need to change when we later add support for persisting the controller's state.

---

With all that said, this does seem to work now, and I think enough time has been spent on this.  I'm going to merge this as it is, but it will likely need revisiting if we continue working on Flight Scheduler.

Thanks for your work on this @WilliamMcCumstie.

WilliamMcCumstie commented:
Agree with your assessment that something isn't quite right surrounding the `Enumerator`, however I differ on the root cause. IMO the problem is the `FifoScheduler` is preforming two very distinctly different roles:
* Scheduling `pending` jobs/tasks and transitioning them to the pseudo-state "allocated", and
* Holding previously "allocated tasks".

This causes also sorts of edge cases concerning cancellation of jobs and their removal from the scheduler. It also means every future implementation of the scheduler has to preform both roles.

Instead the "holding of jobs/task" could be split off into a "JobRegistry" which tracks the state of all known jobs/tasks and manages their persistence. It wouldn't implement a `remove_job` method per se, instead it would quietly remove/backup/archive/etc the `CANCELLED`/`COMPLETED`/`FAILED` jobs. The exact final behaviour could be determine by a plugin.

This frees up the `FifoScheduler` to focus on its primary role, transition jobs from a `pending` to "allocated" pseudo-state. Once this has happened the scheduler can forget about the job entirely, it is no longer responsible for what happens to it after this point. It would likely integrate with the "JobRegistry" for the current state of `pending` jobs (and `running/allocated` if required).

Once we have a better distinction between these roles, the future of the `Enumerator` should be come clear.

benarmston commented:
Thanks for you feedback.  You're right about the need for a job registry of some kind and that the `FifoScheduler` is being used as a pseudo job registry.  You're ideas of how that work sound about right too.

However, I think those points are in addition to rather than in replacement to the points I've noted.

WilliamMcCumstie commented:
:+1: 

Exact implementation TBD
